docs/contributing.rst

```diff
@@ -161,37 +171,75 @@ you can delete it by doing ``git branch -D simulation-enhancements``.
 Obviously this isn't a complete guide to git, but hopefully it
 jump-starts the git learning process.
 
+Testing
+=======
+
+SNCosmo uses pytest to check that all of the code is running as expected. When
+you add new functionality to SNCosmo, you should write a test for that
+functionality. All of the tests can be found in the ``sncosmo/tests``
+directory.
+
+When a new PR is created, the testsuite will be run automatically on a range
+of different machines and conditions using `tox`. You can run these same tests
+locally using ``tox``. First, install `tox`::
+
+      pip install tox
+
+From within the SNCosmo directory, run the test suite::
+
+      tox -e py3
+
+The previous command will run the core test suite with the currently installed
+version of Python. You can run the full test suite with all of the optional
+dependencies by adding the ``-alldeps`` tag::
+
+      tox -e py3-alldeps
+
+Running the tests with the ``-cov`` tag will generate a coverage report::
+
+      tox -e py3-cov
+
+``tox`` can also be used to check the code style::
+
+      tox -e codestyle
+
+or to build the documentation::
+
+      tox -e build_docs
+
+``tox`` uses virtual environments for testing which can be somewhat slow. You
+can alternatively run the test in your own Python environment. First, install
+all of the testing dependencies from the ``test`` section of ``setup.cfg``.
+This can be done automatically when installing SNCosmo with the following
+command::
+
+      pip install -e .[test]
+
+The tests can then be run with the following command::
+
+      pytest --pyargs sncosmo
+
 
 Developer's documentation: release procedure
 ============================================
 
-These are notes mainly for the one person that manages releases.
-Yes, this could be more automated, but it isn't done very often,
-and involves some human verification.
+The release procedure is automated through GitHub Actions. To create a new
+release:
 
 - Update ``docs/history.rst`` with a summary of the new version's changes.
-- Bump version in ``sncosmo/__init__.py``.
```

Comment:
Is this not needed anymore? If not, when/where is this automated?

Reply:
I think I am starting to understand once I read `__init__.py`. Maybe we should comment on how this is now automated, since there are many systems/tools to bump version numbers?

Reply:
The version is now managed by `setuptools_scm`. That package reads the git history and pulls the version from there directly. This approach seems to be the recommended one these days, and is used by astropy. With `setuptools_scm`, when you make a new release on github, it automatically picks it up and updates the version. Any development builds after that also automatically get tagged with the latest version and a dev string afterwards that includes the git commit.

The remaining challenge is how to handle the `__version__` label. There is an `importlib.metadata` package in the standard library in Python 3.8+ that can automatically pull the version from `setuptools_scm`, and this is the way of the future. However, we still want to maintain compatibility with Python 3.6/3.7. There are a few options for this:
1. There is a backport of `importlib_metadata` on PyPI that we could add as a dependency.
2. There is a function in `setuptools` that can handle this, but it then adds a runtime dependency on `setuptools` that adds a bit of extra latency. This is what I decided to do.
3. You can make `setuptools_scm` write out the version to a file on installation that can then be read at runtime. This is what `astropy` does, but it makes things much more complicated.

There are a lot of nice benefits to using the `setuptools_scm` package, and I think that it is the way of the future, but it is a little tricky to use it for Python 3.6/3.7.

I'm not opposed to dropping this and going back to the old hardcoded version if you think that's easier.

Reply:
I like the idea of `setuptools_scm`. More comments in the main tread.